feat(konvoy): add image builder azure docs (#4203)
mesosphere/dcos-docs-site: pages/dkp/konvoy/2.2/image-builder/create-azure-image/index.md

---
layout: layout.pug
navigationTitle: Create Azure Image
title: Create a custom Azure Image
excerpt: Learn how to build a custom Azure Image for use with Konvoy
beta: false
enterprise: false
menuWeight: 60
---

> This procedure describes how to use the Konvoy Image Builder to create a [Cluster API](https://cluster-api.sigs.k8s.io/) compliant Azure Virtual Machine (VM) Image that contains the base operating system you specify as well as all the necessary Kubernetes components.   The Konvoy Image Builder uses variable [`overrides`][overrides] to specify base image and container images to use in your new Azure VM image.

Reviewer: ```suggestion
This procedure describes how to use the Konvoy Image Builder to create a [Cluster API](https://cluster-api.sigs.k8s.io/) compliant Azure Virtual Machine (VM) Image. The VM Image contains the base operating system you specify and all the necessary Kubernetes components. The Konvoy Image Builder uses variable [`overrides`][overrides] to specify base images and container images to use in your new Azure VM Image.
```

Author: Added, with the exception of the `base images`, changed that to `the base image`, since you can only have one base image.

## Prerequisites

Before you begin, you must:

-   Download the latest [Konvoy Image Builder](https://github.com/mesosphere/konvoy-image-builder/releases) bundle (prefixed with `konvoy-image-bundle`) for your OS. Do not use the release prefixed with `konvoy-image-builder`.

-   Create a working `Docker` setup.

## Extract the bundle

Extract the bundle and `cd` into the extracted `konvoy-image-bundle-$VERSION_$OS` folder. The bundled version of `konvoy-image` contains an embedded `docker` image that contains all the requirements for building.

<p class="message--note"><strong>NOTE: </strong> The <code>konvoy-image</code> binary and all supporting folders are also extracted. When run, <code>konvoy-image</code> bind mounts the current working directory (<code>${PWD}</code>) into the container to be used.</p>

> ## Configure Azure prerequisites

Reviewer 2: Markdown needs a blank line before a heading:

```suggestion

## Configure Azure prerequisites
```

> 1.  Log in to Azure:

Reviewer: ```suggestion
1.  Sign in to Azure:
```

    ```sh
    {
      "client_id": "7654321a-1a23-567b-b789-0987b6543a21",
      "client_secret": "Z79yVstq_E.R0R7RUUck718vEHSuyhAB0C",
      "tenant_id": "a1234567-b132-1234-1a11-1234a5678b90"
    }
    ```
-   Set the `AZURE_CLIENT_SECRET` environment variable:

    ```bash
    export AZURE_CLIENT_SECRET
    ```

-   Ensure you have an [override file](../override-files) to configure specific attributes of your Azure image.

## Build the image

Run the `konvoy-image` command to build and validate the image.

```sh
konvoy-image build azure --client-id <azure_client_id> --tenant-id <azure_tenant_id> --overrides override-source-image.yaml --overrides override-images.yaml images/azure/centos-7.yaml
```

By default, the image builder builds in the `westus2` location. To specify another location set the `--location` flag:

```sh
konvoy-image build azure --client-id <azure_client_id> --tenant-id <azure_tenant_id> --location eastus --overrides override-source-image.yaml --overrides override-images.yaml images/ami/centos-7.yaml
```

When the command is complete, the image id is printed and written to `./manifest.json`.    You should then specify this image id when creating the cluster.

> ## Image Gallery

Reviewer: ```suggestion
## Image gallery
```

---
layout: layout.pug
navigationTitle: Create Azure Image
title: Create a custom Azure Image
excerpt: Learn how to build a custom Azure Image for use with Konvoy
beta: false
enterprise: false
menuWeight: 60
---

This procedure describes how to use the Konvoy Image Builder to create a [Cluster API](https://cluster-api.sigs.k8s.io/) compliant Azure Image. Azure images contain configuration information and software to create a specific, pre-configured, operating environment. For example, you can create an Azure image of your current computer system settings and software. The Azure image can then be replicated and distributed, creating your computer system for other users. The Konvoy Image Builder uses variable [`overrides`][overrides] to specify base image and container images to use in your new Azure image.

## Prerequisites

Before you begin, you must:

-   Download the latest [Konvoy Image Builder](https://github.com/mesosphere/konvoy-image-builder/releases) bundle (prefixed with `konvoy-image-bundle`) for your OS. Do not use the release prefixed with `konvoy-image-builder`.

-   Create a working `Docker` setup.

> ## Extract Bundle

Reviewer 3: ```suggestion
## Extract the bundle
```

Extract the bundle and `cd` into the extracted `konvoy-image-bundle-$VERSION_$OS` folder. The bundled version of `konvoy-image` contains an embedded `docker` image that contains all the requirements for building.

<p class="message--note"><strong>NOTE: </strong> The <code>konvoy-image</code> binary and all supporting folders are also extracted. When run, <code>konvoy-image</code> bind mounts the current working directory (<code>${PWD}</code>) into the container to be used.</p>

> 

Reviewer 2: Does the user needs to 'az login' and create a Service Principal to perform these actions? (i.e. should they follow the first few steps of the flow listed here: docs.d2iq.com/dkp/konvoy/2.1/choose-infrastructure/azure/quick-start-azure/#configure-azure-prerequisites).  If so:

```suggestion
## Configure Azure prerequisites

1.  Log in to Azure:

    ```bash
    az login
    ```

    ```sh
    [
      {
        "cloudName": "AzureCloud",
        "homeTenantId": "a1234567-b132-1234-1a11-1234a5678b90",
        "id": "b1234567-abcd-11a1-a0a0-1234a5678b90",
        "isDefault": true,
        "managedByTenants": [],
        "name": "Mesosphere Developer Subscription",
        "state": "Enabled",
        "tenantId": "a1234567-b132-1234-1a11-1234a5678b90",
        "user": {
          "name": "user@azuremesosphere.onmicrosoft.com",
          "type": "user"
        }
      }
    ]
    ```

1.  Create an Azure Service Principal (SP) by running the following command:

    <p class="message--note"><strong>NOTE: </strong>If an SP with the name exists, this command will rotate the password.</p>

    ```bash
    az ad sp create-for-rbac --role contributor --name "$(whoami)-konvoy"
    ```

    ```sh
    {
      "appId": "7654321a-1a23-567b-b789-0987b6543a21",
      "displayName": "azure-cli-2021-03-09-23-17-06",
      "name": "http://azure-cli-2021-03-09-23-17-06",
      "password": "Z79yVstq_E.R0R7RUUck718vEHSuyhAB0C",
      "tenant": "a1234567-b132-1234-1a11-1234a5678b90"
    }
    ```
```

Author: Nice, yep I'll also augment the command with `--query "{ client_id: appId, client_secret: password, tenant_id: tenant }"` so that it outputs all the need to set the env vars and the cli aparams

> -   Set environment `AZURE_CLIENT_SECRET` variable:

Reviewer 2: ```suggestion
-   Set the `AZURE_CLIENT_SECRET` environment variable:
```

    ```bash
    export AZURE_CLIENT_SECRET
    ```

-   Ensure you have an [override file](../override-files) to configure specific attributes of your Azure image.

## Build the image

Run the `konvoy-image` command to build and validate the image.

```sh
konvoy-image build --client-id <azure_client_id> --tenant-id <azure_tenant_id> --overrides override-source-image.yaml --overrides override-images.yaml images/azure/centos-7.yaml
```

> By default it builds in the `westus2` location. To specify another location set the `--location` flag:

Reviewer 2: ```suggestion
By default, the image builder builds in the `westus2` location. To specify another location set the `--location` flag:
```

```

When the command is complete the image id is printed and written to `./manifest.json`.

## Image Gallery

By default Konvoy Image Builder will create a Resource Group, Gallery, and Image Name to store the resulting image in. To specify a specific Resource Group, Gallery, or Image Name flags may be specified:

```sh
--gallery-image-locations location   a list of locatins to publish the image (default same as location)
--gallery-image-name string          the gallery image name to publish the image to
--gallery-image-offer string         the gallery image offer to set (default "dkp")
--gallery-image-publisher string     the gallery image publisher to set (default "dkp")
--gallery-image-sku string           the gallery image sku to set
--gallery-name string                the gallery name to publish the image in (default "dkp")
--resource-group string              the resource group to create the image in (default "dkp")
```

The SKU and Image Name will default to the values found in the image yaml.

> <!--- ## Air Gapped

Reviewer 3: When we get there, this air gapped section needs to be its own topic, not in this file.

Author: totes, removed.
